graphql_compiler/interpreter/debugging.py

```diff
@@ -0,0 +1,186 @@
+from copy import deepcopy
+from dataclasses import dataclass, field
+from typing import Any, ClassVar, Dict, Generic, Iterable, List, Sequence, Tuple, TypeVar
+
+from ..typedefs import Literal
+from .typedefs import DataToken
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _enumerate_starting_at(
+    sequence: Sequence[T],
+    first_index: int,
+) -> Iterable[Tuple[int, T]]:
+    """Enumerate only a portion of a sequence, without slicing the sequence.
+
+    Equivalent to enumerate(sequence[first_index:], start=first_index). The built-in enumerate()
+    always iterates the entire sequence and only changes the counter's value based on
+    the start parameter. In this function, we skip elements with indices in [0, first_index)
+    in the sequence, and only iterate over elements with indices in [first_index, len(sequence)).
+
+    Args:
+        sequence: arbitrary sequence of values over which to iterate
+        first_index: the first index of the sequence to produce in the resulting iterable
+
+    Yields:
+        tuples (index, element_at_that_index_in_sequence), analogous to the built-in enumerate()
+    """
+    for index in range(first_index, len(sequence)):
+        yield index, sequence[index]
+
+
+def _unzip_and_yield_second(iterable: Iterable[Tuple[T, U]]) -> Iterable[U]:
+    """Unpack an iterable of tuples and only yield the second element of each tuple."""
+    for _, second in iterable:
+        yield second
+
+
+@dataclass(frozen=True)
+class AdapterOperation:
+    """The record of an action performed to or by an InterpreterAdapter function as part of a trace.
```

Comment:
what does `an action performed to a function` mean?

Reply:
Yeah this is not my favorite phrasing either. The functions being recorded aren't "plain" functions but generators (AKA coroutines), which may be resumed by calling their `send()` method (in Python, functions can have methods — weird, right?)

This is arguably an action performed _on_ the function, rather than _by_ the function. I probably should have used "on" instead of "to" but I doubt the bad choice of preposition is the crux of the issue.

I had a really hard time figuring out something concise enough to be the top line of this docstring. Any ideas?

Reply:
I'm adding another short paragraph here explaining how input iterable arguments are recorded, which should hopefully clarify this a bit. Let me know if that did the trick, or if I should do something else as well.

Reply:
The `on` definitely helps! I don't have a suggestion for how to make this better so I'd say it is fine :)